### ISSUE TITLE
k3s generated iptables input filter rules are not correct

### ISSUE THREAD

**Issue by v42net** [k3s-io/k3s#3258]
**Environmental Info:**
- k3s version v1.20.6+k3s1 (8d043282)
- go version go1.15.10
 
Node CPU architecture, OS, and Version: 
- Linux kube 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
- Debian version 10.9
- iptables v1.8.2 (legacy)

Cluster Configuration:
- Single k3s install as described in the [Quick-Start Guide](https://rancher.com/docs/k3s/latest/en/quick-start/)

**Describe the bug:**

The `KUBE-ROUTER-INPUT` filter rules `RETURN` to the `INPUT` filter rules when allowing traffic. This will cause the traffic to be dropped when the `INPUT` filter policy is `DROP`. The other k3s generated filter rules correctly use `DROP` or `ACCEPT` for traffic matching those rules.

**Steps To Reproduce:**

1. Install a minimal Debian 10.9 system with curl and iptables (iptables-legacy), but no iptables rules configured. 
2. Install k3s server using the [Quick-Start Guide](https://rancher.com/docs/k3s/latest/en/quick-start/) and confirm it is operating correctly. As a reference for the correct situation use `iptables-save` to save the now active rules as generated by k3s. 
3. Remove k3s with `k3s-uninstall.sh` and reboot or re-install the system to be able to start with a clean slate.
4. Configure iptables rules to allow input from interface `lo`, protocol `icmp`, RELATED and ESTABLISHED traffic, and ssh traffic (tcp/22). Configure the default input and forward policies to drop traffic. The default output policy remains unchanged and accepts traffic. 
5. Configure additional iptables rules based on the *Networking* section of the [Installation Requirements](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/), using the external IP address of the system as the source address (when installing a cluster these rules will be duplicated with the address of each cluster node as the source address).
6. Activate the configured iptables rules and install k3s server using the same procedure as before. Observe that k3s will *not* start correctly. As a reference for the failing situation use `iptables-save` to save the now active rules (the initial rules plus the k3s generated rules).

I did *not* check if kubernetes behaves exactly the same due to lack of resources. In that case this would be a k3s *feature* instead of a k3s bug, until it has been fixed in kubernetes ...

**Expected behavior:**

The `KUBE-ROUTER-INPUT` filter rules should use `-j ACCEPT` when allowing traffic. 

**Actual behavior:**

The `KUBE-ROUTER-INPUT` filter rules `RETURN` to the `INPUT` filter rules when allowing traffic. 

**Additional context / logs:**

The use of iptables-legacy has been configured to prevent warnings in the `k3s check-config` output:
```
# update-alternatives --set iptables /usr/sbin/iptables-legacy
```

The initial iptables rules configured in steps 4 and 5 look like this:
```
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
-A INPUT -m comment --comment "/etc/iptables/rules.v4"
-A INPUT -i lo -j ACCEPT
-A INPUT -p icmp -j ACCEPT
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -p tcp -m state --state NEW -m tcp --dport 22 -j ACCEPT
-A INPUT -p tcp -m state --state NEW -m multiport --dports 80,443 -j ACCEPT
-A INPUT -s 192.168.1.30/32 -p tcp -m state --state NEW -m multiport --dports 2379,2380,6443,10250 -j ACCEPT
-A INPUT -s 192.168.1.30/32 -p udp -m udp --dport 8472 -j ACCEPT
COMMIT
```

To prevent these rules from being reloaded and overwrite the k3s generated rules I'm using the following script:
```
#!/bin/sh
# /etc/network/if-pre-up.d/iptables
if iptables -L | grep -q '/etc/iptables/rules.v4'; then
    echo '/etc/iptables/rules.v4 already loaded'
else
    /sbin/iptables-restore  < /etc/iptables/rules.v4
fi
exit 0
```

As a workaround add the following rules just before the `COMMIT` to the initial iptables rules:
```
-A INPUT -d 10.43.0.0/16 -j ACCEPT
-A INPUT -p tcp -m addrtype --dst-type LOCAL -m multiport --dports 30000:32767 -j ACCEPT
-A INPUT -p udp -m addrtype --dst-type LOCAL -m multiport --dports 30000:32767 -j ACCEPT
```
I am *not* using `iptables-persistent` as that method would also persist the k3s generated rules.

**Comment by brandond** [comment]
The KUBE-ROUTER-* entries come from kube-router's network policy controller, which we have vendored in to K3s. This issue probably belongs over there: https://github.com/cloudnativelabs/kube-router

If you want to disable this component, you can start k3s with --disable-network-policy.

**Comment by v42net** [comment]
Brad, I don't want to disable kube-router as I prefer to use k3s in its most standard way. First as a stand-alone system to get to know k3s and later as a HA k3s cluster. I had a quick look at https://github.com/cloudnativelabs/kube-router, but to raise an issue I need to report the system information:

 - Kube-Router Version (`kube-router --version`): [e.g. 1.0.1]
 - Kube-Router Parameters: [e.g. --run-router --run-service-proxy --enable-overlay --overlay-type=full etc.]
 - Kubernetes Version (`kubectl version`) : [e.g. 1.18.3]
 - Kubernetes Deployment Type: [e.g. EKS, GKE, Kops, Kubeadm, etc.]
 - Kube-Router Deployment Type: [e.g. DaemonSet, System Service]

How do I retrieve this information from k3s so I can raise a valid issue with the kube-router project? Or should a just raise the issue based on the k3s version information? I've got some experience with complex iptables rulesets but a complete newbie in relation to k8s and k3s ...

**Comment by brandond** [comment]
Most of that info isn't relevant since you're using just their network policy controller component that we've built into K3s. I would probably just provide the K3s version you're using.

**Comment by v42net** [comment]
Thanks. I'll have a look at their code myself, so I might even be able to point directly at the relevant lines. I'll provide the link here once I've raised the issue.

**Comment by v42net** [comment]
See  the kube-router issue that causes this issue in k3s: [network_policy_controller.go generates incorrect iptables rules #1076](https://github.com/cloudnativelabs/kube-router/issues/1076)

**Comment by rancher-max** [comment]
I'm going to close this as it has been fixed and confirmed as part of #3373

**Comment by dzegarra** [comment]
after hours of research and trying all suggestions I finally solved my issue by using K3s with Calico.

You can find my solution here: https://dzegarra.hashnode.dev/solution-to-no-internet-access-on-pods-using-k3s-on-armbian
